source/engine/graphics/system/SkeletalAnimationCalculations.cpp

```diff
@@ -0,0 +1,276 @@
+        if (timeInTicks < positionFrames[i+1].timeInTicks)
+        {
+            positionIndex = i;
+            break;
+        }
+    }
+    if (positionIndex == UINT_LEAST32_MAX)
+    {
+        throw NcError("Animation has no position data at time: ", std::to_string(timeInTicks));
+    }
+
+    auto nextPositionIndex = positionIndex + 1;
+    if ((positionIndex + 1) > positionFrames.size())
+    {
+        throw NcError("Animation interpolation index has exceeded the bounds of the frame data.");
+    }
+
+    auto deltaTimeInTicks = positionFrames[nextPositionIndex].timeInTicks - positionFrames[positionIndex].timeInTicks;
+    auto interpolationFactor = (timeInTicks - positionFrames[positionIndex].timeInTicks) / deltaTimeInTicks;
+
+    if (interpolationFactor < 0.0f || interpolationFactor > 1.0f)
+    {
+        throw NcError("Error calculating the interpolation factor: ", std::to_string(interpolationFactor));
+    }
+
+    interpolatedPosition = Lerp(positionFrames[positionIndex].position, positionFrames[nextPositionIndex].position, static_cast<float>(interpolationFactor));
+    return interpolatedPosition;
+}
+
+auto GetInterpolatedRotation(float timeInTicks, const std::vector<nc::asset::RotationFrame> rotationFrames) -> nc::Quaternion
+{
+    auto interpolatedRotation = nc::Quaternion{};
+
+    if (rotationFrames.size() == 0)
+    {
+        throw NcError("Animation has no rotation data.");
+    }
+
+    if (rotationFrames.size() == 1)
+    {
+        return Normalize(rotationFrames[0].rotation);
+    }
+
+    auto rotationIndex = UINT_LEAST32_MAX;
+    for (auto i = 0u; i < rotationFrames.size(); i++)
+    {
+        if (timeInTicks < rotationFrames[i+1].timeInTicks)
+        {
+            rotationIndex = i;
+            break;
+        }
+    }
+    if (rotationIndex == UINT_LEAST32_MAX)
+    {
+        throw NcError("Animation has no rotation data at time: ", std::to_string(timeInTicks));
+    }
+
+    auto nextRotationIndex = rotationIndex + 1;
+    if ((rotationIndex + 1) > rotationFrames.size())
+    {
+        throw NcError("Animation interpolation index has exceeded the bounds of the frame data.");
+    }
+
+    auto deltaTimeInTicks = rotationFrames[nextRotationIndex].timeInTicks - rotationFrames[rotationIndex].timeInTicks;
+    auto interpolationFactor = (timeInTicks - rotationFrames[rotationIndex].timeInTicks) / deltaTimeInTicks;
+
+    if (interpolationFactor < 0.0f || interpolationFactor > 1.0f)
+    {
+        throw NcError("Error calculating the interpolation factor: ", std::to_string(interpolationFactor));
+    }
+
+    interpolatedRotation = Slerp(rotationFrames[rotationIndex].rotation, rotationFrames[nextRotationIndex].rotation, static_cast<float>(interpolationFactor));
+    return Normalize(interpolatedRotation);
+}
+
+auto GetInterpolatedScale(float timeInTicks, const std::vector<nc::asset::ScaleFrame> scaleFrames) -> nc::Vector3
+{
+    auto interpolatedScale = nc::Vector3{};
+    if (scaleFrames.size() == 0)
```

Comment:
Now that I think about it, what do you think about making all the error handling use NC_ASSERT just to compress things a bit. Its not really an assert, as it throws. However, it is something we can turn on and off. I suppose it comes down to whether these are more like sanity checks (prefer NC_ASSERT), or if they're reasonably easy to arrive at(keep as is). Can discuss, if needed.